### PR TITLE
[BE]: Update NCCL submodule to 2.22.3

### PR DESCRIPTION
**Pull request by Skylion007**
Update NCCL submodule to 2.22.3 . This is a pretty major release that has been out for a while and hasn't encountered any issues. It also fixes a lot bugs and enables some new NCCL features: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-22-3.html . It's been out since June so it should be fairly stable.

### CODE REVIEW

**Review by ezyang — DISMISSED** [review]

**Review by nWEIdia — COMMENTED** [review]
The failure reminds me that such updates usually require S3 index updates (nvidia-nccl-cu11 and nvidia-nccl-cu12)

btw, Could this one wait for https://github.com/pytorch/pytorch/pull/132202 please? 

We are targeting a Monday merge, praying for a zero conflict :)

File: .github/scripts/generate_binary_build_matrix.py [hunk 1]
```diff
@@ -51,7 +51,7 @@
         "nvidia-curand-cu11==10.3.0.86; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cusolver-cu11==11.4.1.48; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cusparse-cu11==11.7.5.86; platform_system == 'Linux' and platform_machine == 'x86_64' | "
-        "nvidia-nccl-cu11==2.21.5; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
Comment (atalman) [inline .github/scripts/generate_binary_build_matrix.py:54]: Here https://pypi.org/project/nvidia-nccl-cu11/ version: 2.21.5 is the latest one. I guess we only would need to update cu 12.1 and 12.4

Comment (Skylion007) [reply .github/scripts/generate_binary_build_matrix.py:54]: @eqy Do you know why there are no binaries released for 11.8 here? Is this intentional or a release bug?

Comment (Skylion007) [reply .github/scripts/generate_binary_build_matrix.py:54]: @atalman Unfortunately, this isn't easily possible as Conda use a statically linked one which builds against the git submodule. We could build that one 2.22.5, but it's untested and violates an testing assert. @malfet What are the thoughts going for now that 11.8 is not supported by Nvidia anymore?

Comment (malfet) [reply .github/scripts/generate_binary_build_matrix.py:54]: If 2.22.3 is not supported for 11.8 and we need 2.22.3, we should stop supporting 11.8, but IMO this is a pretty big step to make... On the other hand, those constrains are only needed for poetry, aren't they? In that case, we might build/link with nccl statically for 11.8, but dynamically for 12.x

**Review by malfet — CHANGES_REQUESTED** [review]
Please do not remove `"nvidia-nccl-cu11==2.21.5; platform_system == 'Linux' and platform_machine == 'x86_64' | "` constrains or just add statement, that we don't care about poetry users anymore...

**Comment by Skylion007** [comment]
> Please do not remove `"nvidia-nccl-cu11==2.21.5; platform_system == 'Linux' and platform_machine == 'x86_64' | "` constrains or just add statement, that we don't care about poetry users anymore...

That's not the issue, the issue is we have a divergence of supported NCCL versions. 11.8 only supports up to 2.21.5 while 12 supports newer versions. This will also be a problem for CUDNN especially as we update it to get needed CUDNN attention fixes.

**Comment by Skylion007** [comment]
Yeah, we could try to do some submodule hacking to change the tags in different dockers, but that seems like a pain.

**Comment by kit1980** [comment]
Let's deprecate CUDA 11.8 and update nccl at the same time, for 2.6 release probably.

File: .github/scripts/generate_binary_build_matrix.py (hunk 1 above)
Comment (ngimel) [reply .github/scripts/generate_binary_build_matrix.py:54]: We kinda decided to stop supporting 11.8, so maybe we could update nccl? cc @kwen2501

**Pull request closed without merging**